### PR TITLE
two new modules and improvements to the Leviathan (hard) trial module

### PR DESCRIPTION
**Pull request by CarnifexOptimus**

### CODE REVIEW

File: BossMod/Components/Knockback.cs [hunk 1]
```diff
@@ -137,6 +139,8 @@ protected IEnumerable<(WPos from, WPos to)> CalculateMovements(BossModule module
                     Kind.AwayFromOrigin => from != s.Origin ? (from - s.Origin).Normalized() : new(),
                     Kind.DirForward => s.Direction.ToDirection(),
                     Kind.DirLeft => s.Direction.ToDirection().OrthoL(),
+                    Kind.DirRight => s.Direction.ToDirection().OrthoR(),
+                    Kind.TowardsOrigin => s.Direction.ToDirection().OrthoL().OrthoL(),
```
Comment (awgil) [inline BossMod/Components/Knockback.cs:143]: Should be simply `-s.Direction.ToDirection()`
Also - minor thing - please keep enum and switch order the same.

File: BossMod/Components/Knockback.cs [hunk 2]
```diff
@@ -150,6 +154,12 @@ protected IEnumerable<(WPos from, WPos to)> CalculateMovements(BossModule module
                     break;
             }
         }
+
+        internal void OnCastFinished(BossModule module, Actor caster, ActorCastEvent spell)
+        {
+            throw new NotImplementedException();
```
Comment (awgil) [inline BossMod/Components/Knockback.cs:160]: Why? This is not a good idea, knockback components shouldn't require overriding OnCastFinished

File: BossMod/Components/Knockback.cs [hunk 3]
```diff
@@ -10,8 +10,10 @@ public abstract class Knockback : CastCounter
         public enum Kind
         {
             AwayFromOrigin, // standard knockback - specific distance along ray from origin to target
+            TowardsOrigin, // standard pull - "knockback" to source - forward along source's direction + 180 degrees
```
Comment (awgil) [inline BossMod/Components/Knockback.cs:13]: I feel that properly supporting TowardsOrigin requires a few more changes - primarily min distance - for correctness.
Typically 'attract' effects cap the traveled distance, so that distance between target and origin is never less than a constant >= 0 (and maybe + source/target hitbox radiuses?) In any case, it can't become less than 0 - it would be an 'overshoot', which can't happen with attracts.

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
+using System.IO.Compression;
+using System.Security.Cryptography.X509Certificates;
+using FFXIVClientStructs.FFXIV.Common.Lua;
```
Comment (awgil) [inline BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:3]: That's probably all unneeded :)

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+public enum AID : uint
+{
+    Attack = 872, // Boss->player, no cast, single-target
+    Thundercall = 32212, // Boss->location, 2,5s cast, range 3 circle
+    LightningBolt = 32214, // Boss->self, 3,0s cast, single-target
+    LightningBolt2 = 32215, // BossHelper->location, 3,0s cast, range 6 circle
+    ThunderIV = 32213, // BallOfLevin->self, 7,0s cast, range 18 circle
+    Spark = 32216, // Boss->self, 4,0s cast, range 14-30 donut --> TODO: confirm inner circle size
+    Attack2 = 870, // BossMikros->player, no cast, single-target
```
Comment (awgil) [inline BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:23]: What's BossMikros?

Also by convention autoattacks are called AutoAttack

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    RockThrow = 32217, // BossMikros->location, 3,0s cast, range 6 circle
+    SweepingGouge = 32211, // Boss->player, 5,0s cast, single-target
+};
+
+public enum IconID : uint
+{    
+    AOE = 111, // Thundercall marker
```
Comment (awgil) [inline BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:30]: AOE is a bit too generic, probably Thundercall is a better name

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        public Spark() : base(ActionID.MakeSpell(AID.Spark), new AOEShapeDonut(13,30)) { } 
+    }
+    class SweepingGouge : Components.SingleTargetCast
+    {
+        public SweepingGouge() : base(ActionID.MakeSpell(AID.SweepingGouge)) { }
+    }
+    class Thundercall : Components.LocationTargetedAOEs
+    {
+        private int targeted;
```
Comment (awgil) [inline BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:43]: Should be an `Actor? _target` instead? In general, components shouldn't be player-centric, otherwise features like changing POV mid replay break.

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+        }
+        public override void OnCastStarted(BossModule module, Actor caster, ActorCastInfo spell)
+        {
+            if ((AID)spell.Action.ID == AID.Thundercall)
+                targeted = 0;
+        }
+        public override void AddGlobalHints(BossModule module, GlobalHints hints)
+        {
+             if (targeted > 0)
+                hints.Add("Drop the puddle at the edge of the arena and get away, it will spawn a huge AOE circle");
```
Comment (awgil) [inline BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:59]: These kind of things should be personal hints, rather than global (because it's only relevant to the spell target).
Also I tend to make hints very concise (eg. 'GTFO to the edge'), since you wouldn't really have time to read long full sentences mid fight

File: BossMod/Modules/Endwalker/TreasureHunt/LyssaChrysine.cs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+    {
+        public SkullDasher() : base(ActionID.MakeSpell(AID.SkullDasher)) { }
+    }
+class HeavySmash : Components.StackWithCastTargets
+    {
+        public HeavySmash() : base(ActionID.MakeSpell(AID.HeavySmash), 6) { }
+    }
+class IcePillarSpawn : Components.GenericAOEs
+    {     
+        private float activePillar;
```
Comment (awgil) [inline BossMod/Modules/Endwalker/TreasureHunt/LyssaChrysine.cs:101]: Should be an `Actor? _activePillar`? and then all functions should check if non-null and use it directly, rather than looking up in actor list.

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterH.cs [hunk 10]
```diff
@@ -1,178 +1,10 @@
...
-        }
-    }
+﻿using System.Collections.Generic;
 
+namespace BossMod.Modules.RealmReborn.Trial.T09WhorleaterH;
+ 
     public class T09WhorleaterH : BossModule
     {
-        private List<Actor> _spumes;
+       private List<Actor> _spumes;
```
Comment (awgil) [inline BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterH.cs:7]: This formatting change is unintended?

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterGenericActions.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+    class Hydroshot : PersistentVoidzoneAtCastTarget
+    {
+        public Hydroshot() : base(5, ActionID.MakeSpell(AID.Hydroshot), m => m.Enemies(OID.HydroshotZone), 0) { }
+    }
+    class Dreadstorm : PersistentVoidzoneAtCastTarget
+    {
+        public Dreadstorm() : base(5, ActionID.MakeSpell(AID.Dreadstorm), m => m.Enemies(OID.DreadstormZone), 0) { }
+    }
+    class T09WhorleaterHStates : StateMachineBuilder
```
Comment (awgil) [inline BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterGenericActions.cs:18]: Typically I keep 'simple' components in the main .cs, next to the module, and don't move them out to 'generic actions'. Only if they tend to become big, I then move them to own .cs named after component (if there are several closely related ones - select most important one, and move it + related smaller ones together).

For states - similar, keep in main .cs if trivial, move out to xxxStates.cs if it becomes complex.

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHBodyslam.cs [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        private float Distance;
+        private Angle Direction;
+
+        private float LeviathanZ;
+
+        public override IEnumerable<Source> Sources(BossModule module, int slot, Actor actor)
+        {
+            if (Distance > 0)
+                yield return new(new(), Distance, default, null, Direction, Kind.DirForward);
```
Comment (awgil) [inline BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHBodyslam.cs:16]: First argument should probably be `module.Bounds.Center`?

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHBodyslam.cs [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+            {
+                Distance = 0;
+            }
+        }
+        public override void AddHints(BossModule module, int slot, Actor actor, TextHints hints, MovementHints? movementHints){}
+
+    }
+        class BodySlamAOE : GenericAOEs
+        {
+        private float active;
```
Comment (awgil) [inline BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHBodyslam.cs:49]: Why is it a float?

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs [hunk 14]
```diff
@@ -0,0 +1,57 @@
+using BossMod.Components;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace BossMod.Modules.RealmReborn.Trial.T09WhorleaterH;
+   class SpinningDive : GenericAOEs //TODO: Find out how to detect spinning dives earlier eg. the water column telegraph
+    {     
+        private float activeHelper;
```
Comment (awgil) [inline BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs:8]: Why is it a float? Should be an `Actor?`

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+            if ((AID)spell.Action.ID == AID.SpinningDiveSnapshot)
+            {
+                activeHelper = 0 ;
+            }
+        }
+    }
+
+        class SpinningDiveKB : Knockback //TODO: Find out how to detect spinning dives earlier eg. the water column telegraph
+    {        
+        private float activeHelper;
```
Comment (awgil) [inline BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs:33]: Why is it a float? Should be an `Actor?`

File: BossMod/Modules/Shadowbringers/FATE/Formidable.cs [hunk 16]
```diff
@@ -218,10 +223,48 @@ class SteamDome : Components.KnockbackFromCastTarget
         public SteamDome() : base(ActionID.MakeSpell(AID.SteamDome), 15) { }
     }
 
-    // TODO: stay/move component
     class DynamicSensoryJammer : Components.CastHint
     {
-        public DynamicSensoryJammer() : base(ActionID.MakeSpell(AID.DynamicSensoryJammer), "Stay still!") { }
+        public DynamicSensoryJammer() : base(ActionID.MakeSpell(AID.DynamicSensoryJammer), "") { }
+        private int ec;
```
Comment (awgil) [inline BossMod/Modules/Shadowbringers/FATE/Formidable.cs:229]: Should probably be a BitMask with bit set for each player with corresponding status - so eg. OnStatusGain would do `ec.Set(module.Raid.FindSlot(actor.InstanceID))`

File: BossMod/Modules/Shadowbringers/FATE/Formidable.cs [hunk 17]
```diff
@@ -218,10 +223,48 @@ class SteamDome : Components.KnockbackFromCastTarget
...
+        private int casting;
```
Comment (awgil) [inline BossMod/Modules/Shadowbringers/FATE/Formidable.cs:230]: Why is it an int? Should be a bool?

File: BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+        {
+            if (module.PrimaryActor.CastInfo == null || !module.PrimaryActor.CastInfo.IsSpell())
+                return;
+            switch ((AID)module.PrimaryActor.CastInfo.Action.ID)
+            {
+                case AID.NumbingNoiseRotation:
+                    _RotationDir = module.PrimaryActor.Rotation;
+                    break;
+                case AID.TailSnapRotation:
+                    _RotationDir = module.PrimaryActor.Rotation+180.Degrees();
```
Comment (awgil) [inline BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs:62]: Why do you do all this in Update, rather than in OnCastStarted?

File: BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+            }
+        }
+
+        public override void DrawArenaBackground(BossModule module, int pcSlot, Actor pc, MiniArena arena)
+        {
+            base.DrawArenaBackground(module, pcSlot, pc, arena);
+            if (_remainingHits > 1 && _RotationDirIncrement.Rad != MathF.PI && _numbingnoise == 1)
+                arena.ZoneCone(module.PrimaryActor.Position, 0, _shapeNumbingNoise.Radius, _RotationDir + _RotationDirIncrement, 60.Degrees(), ArenaColor.Danger);
+            if (_remainingHits > 1 && _RotationDirIncrement.Rad != MathF.PI && _tailsnap == 1)
+                arena.ZoneCone(module.PrimaryActor.Position, 0, _shapeTailSnap.Radius, _RotationDir + _RotationDirIncrement, 60.Degrees(), ArenaColor.Danger);
```
Comment (awgil) [inline BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs:73]: Why do you draw this manually, rather than by reporting these aoes in ActiveAOEs?

**Review by awgil — CHANGES_REQUESTED** [review]

File: BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs (hunk 19 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs:73]: because you did it this way in the Burfurlur module

Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs:73]: Gurangatch module i mean

File: BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs (hunk 18 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs:62]: because you did this in the Gurangatch module and I look at existing solutions

File: BossMod/Modules/Shadowbringers/FATE/Formidable.cs (hunk 17 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/FATE/Formidable.cs:230]: why does it matter if i use 0 or 1 or true or false?

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs (hunk 15 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs:33]: i guess i could use int or bool instead of float, but it isn't an actor

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs (hunk 5 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:23]: The adds are called Gymnasiou Pithekos Mikros, I couldn't think of a better name

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs (hunk 4 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:3]: I don't know why VS Code added that

File: BossMod/Components/Knockback.cs (hunk 3 above)
Comment (CarnifexOptimus) [reply BossMod/Components/Knockback.cs:13]: yes, but i needed this kind towards origin for my sugaar module

Comment (CarnifexOptimus) [reply BossMod/Components/Knockback.cs:13]: i manually made the calculations there for distance to boss and actual pull distance

File: BossMod/Components/Knockback.cs (hunk 2 above)
Comment (CarnifexOptimus) [reply BossMod/Components/Knockback.cs:160]: I don't know, I haven't added this. VS Code must have added it on its own

File: BossMod/Components/Knockback.cs (hunk 1 above)
Comment (CarnifexOptimus) [reply BossMod/Components/Knockback.cs:143]: I guess I can change that

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs (hunk 8 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:59]: well i made a target check (if (targeted > 0), thought that would be fine. but i can change it

File: BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs (hunk 6 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Endwalker/TreasureHunt/GymnasiouPithekos.cs:30]: ok will be changed

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHBodyslam.cs (hunk 12 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHBodyslam.cs:16]: I don't know. This part was written by Taurenkey and is only marked as new since I split it into multiple files.
And I am not sure if the attack is originating from center or from opposite edge of the arena. But the code worked in my tests.

File: BossMod/Modules/Shadowbringers/FATE/Formidable.cs (hunk 16 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/FATE/Formidable.cs:229]: i dont know, it works as it is in replay files

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHBodyslam.cs (hunk 13 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHBodyslam.cs:49]: changed it

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs [hunk 20]
```diff
@@ -5,52 +5,52 @@
...
     {        
-        private float activeHelper;       
+        private bool activeHelper;       
          public SpinningDiveKB() : base(ActionID.MakeSpell(AID.SpinningDiveEffect)) { }
         private static AOEShapeRect rect = new(46, 8);
         public override IEnumerable<Source> Sources(BossModule module, int slot, Actor actor)
         {
                 var helper = module.Enemies(OID.SpinningDiveHelper).FirstOrDefault();  
-                if (helper != null && activeHelper >=1)
+                if (helper != null && activeHelper == true)
```
Comment (awgil) [inline BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs:39]: Minor thing, but `if (mybool == true)` is redundant, you can just use `if (mybool)`

File: BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs (hunk 15 above)
Comment (awgil) [reply BossMod/Modules/RealmReborn/Trial/T09WhorleaterH/T09WhorleaterHSpinningDive.cs]: Why isn't it? You set it in the `OnActorCreated`, why don't you save the actor that was created into a variable and use that for drawing/hints later?

File: BossMod/Modules/Shadowbringers/FATE/Formidable.cs (hunk 17 above)
Comment (awgil) [reply BossMod/Modules/Shadowbringers/FATE/Formidable.cs:230]: Because that's the point of boolean types? While you can use integer as a boolean technically, that implies a different meaning (some counter or something). What would a value of 2 mean here etc?

File: BossMod/Modules/Shadowbringers/Foray/Duel/Duel2Lyon/Duel2Lyon.cs [hunk 21]
```diff
@@ -0,0 +1,17 @@
+namespace BossMod.Shadowbringers.Foray.Duel.Duel2Lyon
```
Comment (awgil) [inline BossMod/Modules/Shadowbringers/Foray/Duel/Duel2Lyon/Duel2Lyon.cs:1]: Please move that to a different PR, since it's a work in progress - let's focus on what is already done here.

File: BossMod/Modules/Shadowbringers/FATE/Formidable.cs (hunk 17 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/FATE/Formidable.cs:230]: ok, I changed it to bool

File: BossMod/Modules/Shadowbringers/Foray/Duel/Duel2Lyon/Duel2Lyon.cs (hunk 21 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/Foray/Duel/Duel2Lyon/Duel2Lyon.cs]: done

File: BossMod/Modules/Shadowbringers/FATE/Formidable.cs (hunk 16 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/FATE/Formidable.cs]: implemented it

File: BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs (hunk 18 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs:62]: but I changed it

File: BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs (hunk 19 above)
Comment (CarnifexOptimus) [reply BossMod/Modules/Shadowbringers/HuntA/Sugaar.cs]: changed it as requested
